### ISSUE TITLE
Support starting multiple servers in global setup [jest-dev-server]

### ISSUE THREAD

**Issue by mhipszki** [smooth-code/jest-puppeteer#94]
The current implementation of `jest-dev-server` only allows starting up a single server.

In my use-case I'd like to start a _mock server_ to serve data i.e. mocking out a domain service / API and also like to start the actual _application_ to be tested e.g.:

```javascript
const { setup: setupPuppeteer } = require('jest-environment-puppeteer');
const { setup: setupDevServer } = require('jest-dev-server');

module.exports = async function globalSetup() {
  await setupPuppeteer();

  // start mock server and wait for it to be ready
  await setupDevServer({
    command: `yarn mock-server --port=4000`,
    launchTimeout: 50000,
    port: 4000,
    debug: true,
  });

  // start application and wait for it to be ready
  await setupDevServer({
    command: `yarn start --port=3000`,
    launchTimeout: 50000,
    port: 3000,
    debug: true,
  });
};
```

This would allow one to remove the external dependency (domain service / API) when running tests so that the data returned from the mock server would be deterministic rather than relying on unpredictable network, data etc.

Also, in the global teardown each server would be stopped individually e.g. referring to the port number it's running on:

```javascript
const { teardown: teardownPuppeteer } = require('jest-environment-puppeteer');
const { teardown: teardownDevServer } = require('jest-dev-server')

module.exports = async function globalTeardown() {
  // stop application
  await teardownDevServer({ port: 3000 })

  // stop mock server
  await teardownDevServer({ port: 4000 })

  await teardownPuppeteer();
};
```

`jest-dev-server/src/global.js` could be changed slightly so that each server ref would be stored separately e.g. mapped to a port number allowing us to find and stop it.

Currently `server` is a module level variable and gets overridden when we try to start another server:

https://github.com/smooth-code/jest-puppeteer/blob/master/packages/jest-dev-server/src/global.js#L39

Basically replacing `server` with a hashmap would probably just do it e.g.

```javascript
const servers = {};
// vs
let server;
```

and when a new server is spawned at https://github.com/smooth-code/jest-puppeteer/blob/master/packages/jest-dev-server/src/global.js#L63

```javascript
servers[config.port] = spawn(
  // ...
);
```

and in teardown:

```javascript
export async function teardown({ port }) {
  const server = servers[port]
  if (server) await server.destroy()
}
```

Happy to create a quick PR for this one.

**Comment by mhipszki** [comment]
I guess the `jest-environment-puppeteer` would need to be adapted to these changes as well in https://github.com/smooth-code/jest-puppeteer/blob/master/packages/jest-environment-puppeteer/src/global.js

**Comment by neoziro** [comment]
Hi, I think it is a good idea. The simplest solution seems to be able to specify an array of object or a single object into `server` section of configuration. It will not be a breaking change but just a feature.

You are welcome to propose a PR for that! Be careful of updating all readmes :

- https://github.com/smooth-code/jest-puppeteer/tree/master/packages/jest-dev-server
- https://github.com/smooth-code/jest-puppeteer/tree/master/packages/jest-environment-puppeteer
- https://github.com/smooth-code/jest-puppeteer

Thanks!

**Comment by mhipszki** [comment]
cool, PR to land tonight :)

**Comment by Dartv** [comment]
Progress on this? I need this feature too.

**Comment by mhipszki** [comment]
hiya @Dartv I have time this wee to work on this and have a working PoC locally, will issue a PR soon

**Comment by tybro0103** [comment]
This doesn't use `jest-puppeteer`, but may be handy:

```es6
// test-utils/server.js
const http = require('http');
const app = require('my-app/').default; // an instance of express()
let server;

const start = (port) => {
    app.set('port', port);
    server = http.createServer(app);
    return new Promise((resolve, reject) => {
        server.on('error', reject);
        server.listen(port, () => {
            console.log('------- SERVER LISTENING ON PORT', port);
            resolve();
        });
    });
};

const stop = () => {
    return new Promise((resolve, reject) => {
        server.close((error) => {
            error ? reject(error) : resolve();
        });
    });
};

module.exports = {
    start,
    stop,
};
```

And in my tests:
```es6
import server from './test-utils/server';
const port = 7332;

describe('foo', async () => {
    beforeAll(async () => {
        await server.start(port);
    });
    afterAll(async () => {
        await server.stop();
    });
    it('does thing A', async () => {
        await page.goto(`http://localhost:${port}/`, {waitUntil: 'networkidle2'});
    });
});
```

A big difference here is that the server is running in the same process as the tests. For me that was beneficial so I could mock out the requests to other services my server makes on a per-test basis.

**Comment by pvoznyuk** [comment]
Hey guys, any progress on this PR?

**Comment by neoziro** [comment]
Follow up #134
